Modules/mmapmodule.c

```diff
@@ -1355,16 +1355,19 @@ new_mmap_object(PyTypeObject *type, PyObject *args, PyObject *kwdict)
     m_obj->data = mmap(NULL, map_size, prot, flags, fd, offset);
     Py_END_ALLOW_THREADS
 
-    if (devzero != -1) {
-        close(devzero);
-    }
-
     if (m_obj->data == (char *)-1) {
         m_obj->data = NULL;
-        Py_DECREF(m_obj);
+        /* Should be called immediately after function which set errno. */
         PyErr_SetFromErrno(PyExc_OSError);
+        if (devzero != -1) {
+            close(devzero);
+        }
```

Comment:
Copying these 3 lines sounds overkill. The code would look better if you would just save/restore errno.